### PR TITLE
Extending support to full uint64_t/int64_t range - issue #178 (Integer conversion to unsigned)

### PR DESCRIPTION
**Pull request by twelsby**
This pull request is to address issues raised in issue #178 related to the use of `uint64_t` and `int64_t` types and the inaccuracies involved in parsing, or creating objects with the full range of these values.  Specifically it addresses the following problems:

1. Creating an object with a `uint64_t` value greater than (2<sup>63</sup>) -1 will cause it to be represented as an incorrect negative value internally, and subsequently serialized as the negative value.
2. Numbers are parsed as a `double` initially and then converted to integers if possible, so parsing a number greater than (2<sup>53</sup>) - 1 or less than -(2<sup>53</sup>) + 1 will result in loss of precision.  While this behavior is actually endorsed by RFC7159, due to the common use of `uint64_t`/`int64_t` types in modern software it would be preferable if the full range could be supported with full precision.

The pull request contains extensive changes to the code base however the majority of these are merely to introduce a new type, `number_unsigned_t`, which is required to support the upper range of `uint64_t`/`int64_t` values.  The flow on effect of this is the addition of companion constructors, operators and miscellaneous code that is required to implement the new type.  There are also very substantial changes to the unit testing (now just shy of 8000 assertions), again to accommodate the new type.  This new code is in the same form as that for the existing types so there will be no surprises.  A small number of additional unit tests check the performance of the parsing against the full range of the `uint64_t`/`int64_t` types.

The most controversial change is the change to the number parsing function `get_number()`.  This now attempts to parse the number both as an integer using `std::strloull()`/`std::strtoll()` and also using `std::strtold()` and then stores using the best representation.  This method was presented in the discussion for issue [#178](https://github.com/nlohmann/json/issues/178).

This method is far from ideal however the most likely alternatives seem to be either writing some custom (non-standard library based) number parsing function, or attempting to scan ahead in the string before parsing, or modifying the lexer to make the determination, or just accepting the reduced range.  Neither of these alternatives seems ideal.

### CODE REVIEW

**Comment by nlohmann** [comment]
Hi @twelsby, thanks for your work! I'll have a look at the diff. In the meantime, could you check why the unit tests fail on Travis?

File: src/json.hpp.re2c [hunk 1]
```diff
@@ -2027,20 +2197,43 @@ class basic_json
...
+    {
+        return m_type == value_t::number_unsigned;
     }
 
     /*!
     @brief return whether value is a floating-point number
 
-    This function returns true iff the JSON value is a floating-point number.
-    This excludes integer values.
+    This function returns true if the JSON value is a floating-point number.
```
Comment (nlohmann) [inline src/json.hpp.re2c]: In fact, I used "iff" as (common?) abbreviation for "if and only iff" here.

Comment (twelsby) [inline src/json.hpp.re2c]: Sorry I didn't know that one and thought it was a typo.

File: src/json.hpp.re2c [hunk 2]
```diff
@@ -618,7 +700,8 @@ class basic_json
         boolean,        ///< boolean value
         number_integer, ///< number value (integer)
         number_float,   ///< number value (floating-point)
-        discarded       ///< discarded by the the parser callback function
+        discarded,      ///< discarded by the the parser callback function
+        number_unsigned ///< number value (unsigned integer)
```
Comment (nlohmann) [inline src/json.hpp.re2c]: Could you please move `number_unsigned` behind `number_integer` (with the required adjustment in the ordering), just to have the numbers grouped together?

File: src/json.hpp.re2c [hunk 3]
```diff
@@ -6906,17 +7178,40 @@ class basic_json
 
         @throw std::range_error if passed value is out of range
         */
-        long double get_number() const
+        void get_number(basic_json& result) const
```
Comment (nlohmann) [inline src/json.hpp.re2c]: In the documentation, there is still the `@return` part which makes no sense any more as the function is `void`. Could you please adjust the documentation to described the side effect of the function?

File: src/json.hpp.re2c [hunk 4]
```diff
@@ -6906,17 +7178,40 @@ class basic_json
...
 
-            // return float_val if the whole number was translated and NAN
-            // otherwise
-            return (reinterpret_cast<lexer_char_t*>(endptr) == m_cursor) ? float_val : NAN;
+            // Parse it as a double
+            const auto float_val = strtold(reinterpret_cast<typename string_t::const_pointer>(m_start),&endptr);
+            long double int_part;
+            const auto frac_part = std::modf(float_val, &int_part);
+
+            // Test if the double or integer is a better representation
```
Comment (nlohmann) [inline src/json.hpp.re2c]: Could you please add a comment to briefly describe the three parts of the disjunction?

File: src/json.hpp.re2c [hunk 5]
```diff
@@ -4618,14 +4837,16 @@ class basic_json
     */
     friend bool operator<(const value_t lhs, const value_t rhs)
     {
-        static constexpr std::array<uint8_t, 7> order = {{
+        static constexpr std::array<uint8_t, 9> order = {{
```
Comment (nlohmann) [inline src/json.hpp.re2c]: See the comment regarding the order inside the enum above. I'm not sure if still 9 values are required here.

File: src/json.hpp.re2c [hunk 6]
```diff
@@ -1225,14 +1320,72 @@ class basic_json
...
+    @ref number_unsigned_t.
+
+    @param[in] val  an unsigned integer to create a JSON number from
+
+    @complexity Constant.
+
+    @sa @ref basic_json(const number_unsigned_t) -- create a number value
+    (unsigned)
+
+    @since version 1.0.0
```
Comment (nlohmann) [inline src/json.hpp.re2c]: Please make sure to use `@since version 1.0.1` in all functions you added.

File: src/json.hpp.re2c (hunk 2 above)
Comment (twelsby) [inline src/json.hpp.re2c]: I put it at the end because I was concerned that renumbering the enum might break existing code that was expecting certain types to have certain numerical values.  I agree that after number_integer would be a much better position aesthetically. If you are happy that the renumbering won't cause any flow on problems I'll change it.  The added benefit is that it removes the necessity to insert a filler value in the vector that is used for the type ordering comparison (is null < bool etc).

The Travis problem is caused by Valgrind - I'm not sure why but it also happens when I run the unit tests locally under Valgrind.  For some reason Valgrind is truncating the parsed value.  Seems rather bizarre but I'm looking into it.

**Comment by nlohmann** [comment]
I looked at the diffs and adde some comments. Wow - I am impressed how much effort you put to extend all the unit tests!

File: src/json.hpp.re2c (hunk 2 above)
Comment (nlohmann) [inline src/json.hpp.re2c]: Hm. The values of the enum are actually available via public member functions. As this may break existing code, then the extension should be in the 2.0.0 release. I'd rather have clean code than something ugly just to have this edge case covered by a 1.x.x release.

About the tests: it would be great to convince Valgrind to accept the code. If this is some kind of bug, then you give these tests a label so they can be skipped when run with Valgrind...

Comment (twelsby) [inline src/json.hpp.re2c]: Ok, I'll change the order and add from 2.0.0 to the changes.

I've solved the Valgrind problem.  It was down to documented Valgrind behaviour, which is unable to deal with floating point calculations involving `long double` and as a result it was truncating them to `double` and causing the `if(int_part != m_value.number_unsigned)` test to return true which caused the result to be stored as a `number_float_t` instead of a `number_unsigned_t`. For some reason this only happened with the unsigned case... but in any event it doesn't matter, the solution is simply to parse with `strtod` instead of `strtold`, which I should have been doing anyway.  Introducing a `long double` just introduces compiler dependant behaviour (because VS treats them as `double` by default).

While debugging that issue I came up with a better way of parsing which I am still exploring.  It  only requires one parse for integers and has a much simpler (and importantly more obvious) decision methodology.  I would have had it up and running already only I made the fateful decision to upgrade my Ubuntu installation...

**Comment by Furcube** [comment]
Awesome amount of changes.

A little addition: you should set errno = 0 before calling strto<something> functions.
From linux man page for a strtoull:
```
NOTES
       Since strtoul() can legitimately return 0 or LONG_MAX (LLONG_MAX for strtoull()) on both success and failure, the calling program should set errno to 0 before the call, and then determine if an  error
       occurred by checking whether errno has a non-zero value after the call.
```

**Comment by twelsby** [comment]
Good pickup @Furcube I completely forgot about that.  I'll fix it.

**Comment by Furcube** [comment]
result.m_value.number_integer = std::strtoll

This will be very bad if NumberIntegerType is not a int64_t, but int32_t for example
Need to change conversion function to appropriate or do range check additionaly to ERANGE.

**Comment by twelsby** [comment]
@Furcube, another good point.   I think the best way would be to store it in a temporary `int64_t`` /`uint64_t`, check the range of `NumberIntegerType` and cast to either `NumberIntegerType` or `NumberFloatType` as appropriate.  One extra `if`/`else`, so not a huge problem.

**Comment by twelsby** [comment]
The latest commit has the following changes:

1. Fixed the errno issue identified by @Furcube 
2. Fixed the issue with small integers identified by @Furcube 
3. Cleaned things up a bit to increase code reuse using a helper function
4. Implemented the fix referred to in issue #167 - AppVeyor now passes
5. Removed the use of `approx` as discussed in issue #185.  Understandably this will cause angst for some but as I stated in the issue it isn't doing what it is designed for and what it is designed for shouldn't be done.  I look forward to comments on the issue and have no issues continuing with the current way of doing things if any concerns cannot be placated.

**Comment by twelsby** [comment]
Latest build adds a fix for #171 by adding two extra template overloads for operator[] for T* arguments.  Seems to work well with no apparent side effects.

File: src/json.hpp.re2c [hunk 7]
```diff
@@ -65,6 +65,12 @@ Class @ref nlohmann::basic_json is a good entry point for the documentation.
     #endif
 #endif
 
+// disable float-equal warnings on GCC/clang
+#if defined(__clang__) || defined(__GNUC__) || defined(__GNUG__)
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wfloat-equal"
+#endif
+
```
Comment (nlohmann) [inline src/json.hpp.re2c]: I think this part with the `#ifdef`s and `#pragma`s is really ugly... I am not 100% convinced yet to have something like this in the library...

**Comment by nlohmann** [comment]
Hi @twelsby, could you please check #186 with your `get_number` function?

File: src/json.hpp.re2c (hunk 7 above)
Comment (twelsby) [inline src/json.hpp.re2c]: This is necessary to suppress floating point warnings under gcc/clang.  The only real alternatives are to:

1. Accept the warnings - they are only warnings after all (ugly though)
2. Change the build settings in Travis.yml so that -Wfloat-equal is omitted at the command line - this works for us but if the user decides to compile with the warning, and doesn't understand floating point, then they may come to the conclusion that this is a bug in our code
3. Add the #pragma without the conditional block - this creates its own warning about unknown #pragma under VS2015
4. Reinstate approx()
5. Come up with some other means of avoiding the warning if such a thing exists

I will revert this for now and add it to its own pull request as it is really a separate issue.

**Comment by twelsby** [comment]
@nlohmann, can you please close this pull request.  I accidentally created it from my master which is causing all sorts of problems as I would like to point master elsewhere but can't while this is open.  I will create another pull request with just the changes relevant to this subject and add a reference to this one.  Unfortunately github does not allow changing the pull request source branch (although its an open feature request).

**Comment by nlohmann** [comment]
Closed as requested by @twelsby.
